src/sentry/api/endpoints/debug_files.py

```diff
@@ -395,6 +395,215 @@ def post(self, request: Request, project: Project) -> Response:
         return Response({"associatedDsymFiles": []})
 
 
+def get_file_info(files, checksum):
+    """
+    Extracts file information from files given a checksum.
+    """
+    file = files.get(checksum)
+    if file is None:
+        return None
+
+    name = file.get("name")
+    debug_id = file.get("debug_id")
+    chunks = file.get("chunks", [])
+
+    return name, debug_id, chunks
+
+
+def batch_assemble(project, files):
+    """
+    Performs assembling in a batch fashion, issuing queries that span multiple files.
+    """
+    # We build a set of all the checksums that still need checks.
+    checksums_to_check = {checksum for checksum in files.keys()}
+    file_response = {}
+
+    # 1. Exclude all files that have already an assemble status.
+    checksums_with_status = set()
+    for checksum in checksums_to_check:
+        # First, check the cached assemble status. During assembling, a
+        # `ProjectDebugFile` will be created, and we need to prevent a race
+        # condition.
+        state, detail = get_assemble_status(AssembleTask.DIF, project.id, checksum)
+        if state == ChunkFileState.OK:
+            file_response[checksum] = {
+                "state": state,
+                "detail": None,
+                "missingChunks": [],
+                "dif": detail,
+            }
+            checksums_with_status.add(checksum)
+        elif state is not None:
+            file_response[checksum] = {"state": state, "detail": detail, "missingChunks": []}
+            checksums_with_status.add(checksum)
+
+    checksums_to_check -= checksums_with_status
```

Comment:
You could skip these intermediate sets and remove checksums from the whole set as you go. But it's fine either way.

Reply:
You can't do that unfortunately, you are not allowed to remove contents from a collection when iterating it.

Reply:
Oh, duh. A lesson I learned long ago in Rust but somehow forgot when it comes to Python.